From b2fe1c6915983b3653e695fd463d88745d5ab138 Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 12:26:04 -0400
Subject: [ethernet] tg3: update rx_jumbo_pending ring param only when jumbo frames are enabled

Message-id: <1410179173-14304-11-git-send-email-ivecera@redhat.com>
Patchwork-id: 91666
O-Subject: [RHEL7.1 PATCH 10/19] tg3: update rx_jumbo_pending ring param only when jumbo frames are enabled
Bugzilla: 1088637
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

BZ: #1088637

Upstream commit(s):
commit ba67b510035141bd89b40bf65efa0a79834311ca
Author: Ivan Vecera <ivecera@redhat.com>
Date:   Thu Apr 17 14:51:08 2014 +0200

    tg3: update rx_jumbo_pending ring param only when jumbo frames are enabled

    The patch fixes a problem with dropped jumbo frames after usage of
    'ethtool -G ... rx'.

    Scenario:
    1. ip link set eth0 up
    2. ethtool -G eth0 rx N # <- This zeroes rx-jumbo
    3. ip link set mtu 9000 dev eth0

    The ethtool command set rx_jumbo_pending to zero so any received jumbo
    packets are dropped and you need to use 'ethtool -G eth0 rx-jumbo N'
    to workaround the issue.
    The patch changes the logic so rx_jumbo_pending value is changed only if
    jumbo frames are enabled (MTU > 1500).

    Signed-off-by: Ivan Vecera <ivecera@redhat.com>
    Acked-by: Michael Chan <mchan@broadcom.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/broadcom/tg3.c b/drivers/net/ethernet/broadcom/tg3.c
index b8a1621..03389c3 100644
--- a/drivers/net/ethernet/broadcom/tg3.c
+++ b/drivers/net/ethernet/broadcom/tg3.c
@@ -12329,7 +12329,9 @@ static int tg3_set_ringparam(struct net_device *dev, struct ethtool_ringparam *e
  if (tg3_flag(tp, MAX_RXPEND_64) &&
      tp->rx_pending > 63)
   tp->rx_pending = 63;
- tp->rx_jumbo_pending = ering->rx_jumbo_pending;
+
+ if (tg3_flag(tp, JUMBO_RING_ENABLE))
+  tp->rx_jumbo_pending = ering->rx_jumbo_pending;
 
  for (i = 0; i < tp->irq_max; i++)
   tp->napi[i].tx_pending = ering->tx_pending;
-- 
1.7.1